iroha_p2p/src/network.rs

```diff
@@ -0,0 +1,134 @@
+use rand::{rngs::StdRng, seq::SliceRandom, SeedableRng};
+use iroha_crypto::Hash;
+use crate::peer::Peer;
+use crate::peer_state::State;
+use tokio::io;
+use tokio::net::{TcpListener, TcpStream};
+use std::sync::{Arc, RwLock};
+use tokio::io::{AsyncReadExt, AsyncWriteExt};
+use tokio::time::Duration;
+
+pub struct Network {
+    /// Server socket will listen on this address
+    listen_addr: String,
+    /// Current order of peers. The roles of peers are defined based on this order.
+    sorted_peers: Arc<RwLock<Vec<Peer>>>,
+    /// Maximum faulty peers in a network.
+    max_faults: u32,
+    /// Hash of last block in chain
+    last_hash: Option<Hash>
+}
+
+impl Network {
+    pub fn new(listen_addr: &str, peers: Vec<Peer>, max_faults: u32) -> Self {
+        let mut network = Self {
+            listen_addr: listen_addr.to_owned(),
+            sorted_peers: Arc::new(RwLock::new(peers)),
+            max_faults,
+            last_hash: None
+        };
+        network.sort_peers(None, 0);
+        network
+    }
+
+    pub async fn start(&mut self) -> io::Result<()> {
+        self.run_manager_task().await;
+        let listener = TcpListener::bind(&self.listen_addr).await.unwrap();
+        loop {
+            let (mut socket, addr) = listener.accept().await?;
+            let peers = Arc::clone(&self.sorted_peers);
+            tokio::spawn(async move {
+                let mut peer = Peer::new(addr, socket, None, State::Connected);
+                update_peer(&mut peer).await.expect("Unable to read client handshake");
+                peers.write().unwrap().push(peer);
+            }).await?;
+        }
+    }
+
+    pub async fn connect_peer(&mut self, addr: String) -> io::Result<()> {
+        let peers = Arc::clone(&self.sorted_peers);
+        drop(
+            tokio::spawn(async move {
+                let stream = TcpStream::connect(addr).await.expect("Unable to connect to peer");
+                let mut peer = Peer::new(stream.peer_addr().unwrap(), stream, None, State::Connecting);
+                update_peer(&mut peer).await.expect("Unable to send client handshake");
+                peers.write().unwrap().push(peer);
+            }).await
+        );
+        Ok(())
+    }
+
+    pub async fn run_manager_task(&mut self) -> io::Result<()> {
+        let peers = Arc::clone(&self.sorted_peers);
+        let delay = Duration::from_millis(250);
+        drop(
+            tokio::spawn(async move {
+                loop {
+                    tokio::time::sleep(delay);
```

Comment:
`.await` here is forgotten. Download [cargo-lints](https://crates.io/crates/cargo-lints) (we are using it in ci) and just run `cargo lints clippy` on this crate.